content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md

```diff
@@ -0,0 +1,772 @@
+        routeAction:
+          single:
+            upstream:
+              name: echo
+              namespace: gloo-system
+```
+
+We can apply these resources with the following commands:
+
+```bash
+kubectl apply -f https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/1-setup/upstream.yaml
+kubectl apply -f https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/1-setup/vs.yaml
+```
+
+Once we apply these two resources, we can start to send traffic to the application through Gloo:
+
+```bash
+➜ curl $(glooctl proxy url)/
+version:v1
+```
+
+Our setup is complete, and our cluster now looks like this:
+
+![](https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/1-setup/setup.png)
+
+## Two-Phased Rollout Strategy
+
+Now we have a new version `v2` of the echo application that we wish to roll out. We know that when the  
+rollout is complete, we are going to end up with this picture:
+
+![](https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/4-decommissioning-v1/end-state.png)
+
+However, to get there, we may want to perform a few rounds of testing to ensure the new version of the application
+meets certain correctness and/or performance acceptance criteria. In this post, we'll introduce a two-phased approach to 
+canary rollout with Gloo, that could be used to satisfy the vast majority of acceptance tests. 
+
+In the first phase, we'll perform smoke and correctness tests by routing a small segment of the traffic to the new version 
+of the application. In this demo, we'll use a header `stage: canary` to trigger routing to the new service, though in 
+practice it may be desirable to make this decision based on another part of the request, such as a claim in a verified JWT. 
+
+In the second phase, we've already established correctness, so we are ready to shift all of the traffic over to the new 
+version of the application. We'll configure weighted destinations, and shift the traffic while monitoring certain business 
+metrics to ensure the service quality remains at acceptable levels. Once 100% of the traffic is shifted to the new version, 
+the old version can be decommissioned. 
+
+In practice, it may be desirable to only use one of the phases for testing, in which case the other phase can be 
+skipped. 
+
+## Phase 1: Initial canary rollout of v2
+
+In this phase, we'll deploy `v2`, and then use a header `stage: canary` to start routing a small amount of specific 
+traffic to the new version. We'll use this header to perform some basic smoke testing and make sure `v2` is working the 
+way we'd expect:
+
+![](https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/2-initial-subset-routing-to-v2/subset-routing.png)
```

Comment:
1. This image should have an alt attribute.
2. Can you copy the image into the repo so that it's served via the content CDN and the image is in source control?